File: python/aiomemq.py
import sys
import asyncio
import random
import json
import traceback
import logging
from collections import defaultdict, deque
from typing import Any, Callable, Deque, Dict, Optional, Union

logging.basicConfig(level=logging.CRITICAL + 1)

DEFAULT_PORT = 7000
DEFAULT_CACHE_SIZE = 100

topics: Dict[str, set] = defaultdict(set)
topics_reverse: Dict[asyncio.StreamWriter, set] = defaultdict(set)
caches: Dict[str, deque] = defaultdict(lambda: deque(maxlen=DEFAULT_CACHE_SIZE))
indexs: Dict[str, int] = defaultdict(int)

template_subscribe: Dict[str, Dict[str, Union[type, bool]]] = {
    "command": {"type": str, "required": True},
    "topic": {"type": str, "required": True},
    "last_seen": {"type": int, "required": False},
    "cache": {"type": bool, "required": False},
}

template_unsubscribe: Dict[str, Dict[str, Union[type, bool]]] = {
    "command": {"type": str, "required": True},
    "topic": {"type": str, "required": True},
}

template_send: Dict[str, Dict[str, Union[type, bool, list]]] = {
    "command": {"type": str, "required": True},
    "topic": {"type": str, "required": True},
    "msg": {"type": str, "required": True},
    "delivery": {"type": str, "required": True, "values": ["all", "one"]},
    "cache": {"type": bool, "required": False},
}


def template_match(template: Dict[str, Dict[str, Any]], obj: Dict[str, Any]) -> bool:
    if not set(obj.keys()).issubset(template.keys()):
        return False
    for key, props in template.items():
        if props["required"] and key not in obj:
            return False
        if key in obj and not isinstance(obj[key], props["type"]):
            return False
        if "values" in props and obj[key] not in props["values"]:
            return False
    return True


def send_cmd(writer: asyncio.StreamWriter, cmd: Dict[str, Any]) -> None:
    writer.write((json.dumps(cmd) + "\r\n").encode("utf8"))


def send_success(writer: asyncio.StreamWriter) -> None:
    cmd = {"success": True}
    send_cmd(writer, cmd)


def send_failure(writer: asyncio.StreamWriter, reason: str) -> None:
    cmd = {"success": False, "reason": reason}
    send_cmd(writer, cmd)


def send_cached(writer: asyncio.StreamWriter, topic: str, last_seen: int) -> None:
    for cmd in caches[topic]:
        if cmd["index"] > last_seen:
            send_cmd(writer, cmd)
    new_cache: Deque[Dict[str, Any]] = deque(maxlen=cache_size)
    for cmd in caches[topic]:
        if cmd["index"] <= last_seen:
            new_cache.append(cmd)
        else:
            if cmd["delivery"] == "all":
                new_cache.append(cmd)
    caches[topic] = new_cache


def handle_subscribe(cmd: Dict[str, Any], writer: asyncio.StreamWriter) -> None:
    topics[cmd["topic"]].add(writer)
    topics_reverse[writer].add(cmd["topic"])
    last_seen = int(cmd["last_seen"]) if "last_seen" in cmd else -1
    send_success(writer)
    if cmd.get("cache", True):
        send_cached(writer, cmd["topic"], last_seen)


def handle_unsubscribe(cmd: Dict[str, Any], writer: asyncio.StreamWriter) -> None:
    topics[cmd["topic"]].remove(writer)
    topics_reverse[writer].remove(cmd["topic"])
    send_success(writer)


def handle_send(cmd: Dict[str, Any], writer: asyncio.StreamWriter) -> None:
    cmd["index"] = indexs[cmd["topic"]]
    indexs[cmd["topic"]] += 1
    do_cache = cmd.get("cache", True)
    if cmd["delivery"] == "all":
        subscribers = topics[cmd["topic"]]
    else:  # delivery == 'one':
        if len(topics[cmd["topic"]]) == 0:
            subscribers = set([])
        else:
            which = random.randint(0, len(topics[cmd["topic"]]) - 1)
            subscribers = set([list(topics[cmd["topic"]])[which]])
            do_cache = False
    if do_cache:
        caches[cmd["topic"]].append(cmd)
    for subscriber in subscribers:
        send_cmd(subscriber, cmd)
    send_success(writer)


def verify_command(cmd: Dict[str, Any]) -> bool:
    templates: Dict[str, Dict[str, Dict[str, Any]]] = {
        "subscribe": template_subscribe,
        "unsubscribe": template_unsubscribe,
        "send": template_send,
    }
    if "command" not in cmd or not isinstance(cmd["command"], str):
        return False
    template: Optional[Dict[str, Dict[str, Any]]] = templates.get(cmd["command"])
    if template is None or not template_match(template, cmd):
        return False
    return True


def handle_command(cmd: Dict[str, Any], writer: asyncio.StreamWriter) -> None:
    handlers = {
        "subscribe": handle_subscribe,
        "unsubscribe": handle_unsubscribe,
        "send": handle_send,
    }
    handler = handlers.get(cmd["command"])
    if handler:
        handler(cmd, writer)


async def handle_client(
    reader: asyncio.StreamReader, writer: asyncio.StreamWriter
) -> None:
    logging.info("New client connected...")
    line = str()
    try:
        while line.strip() != "quit":
            line = await reader.readline()
            if not line:  # Check for EOF
                logging.info("Client disconnected (EOF)")
                break
            try:
                line = line.decode("utf8").strip()
            except UnicodeDecodeError:
                send_failure(writer, "Could not decode input as UTF-8")
                continue
            if line == "":
                continue
            logging.info(f"Received: {line}")
            try:
                cmd = json.loads(line)
            except json.JSONDecodeError:
                send_failure(writer, "Could not parse json")
            else:
                try:
                    if verify_command(cmd):
                        handle_command(cmd, writer)
                    else:
                        send_failure(writer, "Malformed json message")
                except Exception as e:
                    logging.info(traceback.format_exc())
                    send_failure(writer, "Internal exception")
    except Exception as e:
        logging.info(traceback.format_exc())
    finally:
        if writer in topics_reverse:
            for topic in topics_reverse[writer]:
                topics[topic].remove(writer)
            del topics_reverse[writer]
        writer.close()
        await writer.wait_closed()
        logging.info("Client disconnected...")


async def run_server(host: str, port: int) -> None:
    server = await asyncio.start_server(handle_client, host, port)
    logging.info(f"Listening on {host}:{port}...")
    async with server:
        await server.serve_forever()


if __name__ == "__main__":
    if len(sys.argv) not in [1, 2, 3]:
        logging.info(f"Usage: python3 aiomemq.py <port> <cache_size>")
        logging.info(f"  <port>       - optional, default {DEFAULT_PORT}")
        logging.info(f"  <cache_size> - optional, default {DEFAULT_CACHE_SIZE}")
        sys.exit(1)
    port = int(sys.argv[1]) if len(sys.argv) == 2 else DEFAULT_PORT
    cache_size = int(sys.argv[2]) if len(sys.argv) == 3 else DEFAULT_CACHE_SIZE
    caches = defaultdict(lambda: deque(maxlen=cache_size))
    asyncio.run(run_server(host="localhost", port=port))

File: javascript/aiomemq.js
'use strict';

const net = require('net');
const readline = require('readline');
const { StringDecoder } = require('string_decoder');
const decoder = new StringDecoder('utf8');
const { once } = require('events');
const logging = console;

class DefaultDict {
    constructor(defaultInit) {
        return new Proxy({}, {
            get: (target, name) => name in target ?
                target[name] :
                (target[name] = typeof defaultInit === 'function' ?
                    defaultInit() : // Call the function directly
                    defaultInit)
        });
    }
}

class Deque extends Array {
    constructor(maxlen) {
        super();
        this.maxlen = maxlen;
    }

    push(...args) {
        if (args.length + this.length > this.maxlen) {
            this.splice(0, args.length + this.length - this.maxlen);
        }
        return super.push(...args);
    }
}

function sendCmd(socket, cmd) {
    socket.write(JSON.stringify(cmd) + "\r\n", 'utf8');
}

function sendSuccess(socket) {
    const cmd = { success: true };
    sendCmd(socket, cmd);
}

function sendFailure(socket, reason) {
    const cmd = { success: false, reason: reason };
    sendCmd(socket, cmd);
}

const template_subscribe = {
    command: { type: "string", required: true },
    topic: { type: "string", required: true },
    last_seen: { type: "number", required: false },
    cache: { type: "boolean", required: false }
};

const template_unsubscribe = {
    command: { type: "string", required: true },
    topic: { type: "string", required: true }
};

const template_send = {
    command: { type: "string", required: true },
    topic: { type: "string", required: true },
    msg: { type: "string", required: true },
    delivery: { type: "string", required: true, values: ["all", "one"] },
    cache: { type: "boolean", required: false }
};

function templateMatch(template, obj) {
    const objKeys = new Set(Object.keys(obj));
    const templateKeys = new Set(Object.keys(template));
    
    if (![...objKeys].every(key => templateKeys.has(key))) {
        return false;
    }

    for (const [key, props] of Object.entries(template)) {
        if (props.required && !(key in obj)) {
            return false;
        }
        if (key in obj && typeof obj[key] !== props.type) {
            return false;
        }
        if ('values' in props && !props.values.includes(obj[key])) {
            return false;
        }
    }
    
    return true;
}

function verifyCommand(cmd) {
    const templates = {
        subscribe: template_subscribe,
        unsubscribe: template_unsubscribe,
        send: template_send
    };

    if (!cmd.hasOwnProperty("command") || typeof cmd.command !== "string") {
        return false;
    }

    const template = templates[cmd.command];
    if (!template || !templateMatch(template, cmd)) {
        return false;
    }

    return true;
}

function sendCached(socket, topic, lastSeen) {
    const cache = caches[topic];

    // Send commands from the cache that have an index greater than lastSeen
    for (const cmd of cache) {
        if (cmd.index > lastSeen) {
            sendCmd(socket, cmd);
        }
    }

    // Create a new cache and populate it with the appropriate commands
    const newCache = new Deque(cacheSize);
    for (const cmd of cache) {
        if (cmd.index <= lastSeen) {
            newCache.push(cmd);
        } else if (cmd.delivery === "all") {
            newCache.push(cmd);
        }
    }

    caches[topic] = newCache;
}

function handleSubscribe(cmd, socket) {
    if (!topics[cmd.topic]) {
        topics[cmd.topic] = new Set();
    }
    if (!topics_reverse[socket]) {
        topics_reverse[socket] = new Set();
    }
    
    topics[cmd.topic].add(socket);
    topics_reverse[socket].add(cmd.topic);
    
    const lastSeen = cmd.hasOwnProperty('last_seen') ? parseInt(cmd.last_seen, 10) : -1;

    sendSuccess(socket);

    if (cmd.hasOwnProperty('cache') ? cmd.cache : true) {
        sendCached(socket, cmd.topic, lastSeen);
    }
}

function handleUnsubscribe(cmd, socket) {
    if (topics[cmd.topic]) {
        topics[cmd.topic].delete(socket);
    }
    if (topics_reverse[socket]) {
        topics_reverse[socket].delete(cmd.topic);
    }
    sendSuccess(socket);
}

function handleSend(cmd, socket) {
    cmd.index = indexs[cmd.topic];
    indexs[cmd.topic] += 1;

    let doCache = cmd.hasOwnProperty('cache') ? cmd.cache : true;
    let subscribers;

    if (cmd.delivery === 'all') {
        subscribers = topics[cmd.topic];
    } else {  // delivery === 'one'
        if (topics[cmd.topic].size === 0) {
            subscribers = new Set();
        } else {
            const which = Math.floor(Math.random() * topics[cmd.topic].size);
            subscribers = new Set([Array.from(topics[cmd.topic])[which]]);
            doCache = false;
        }
    }

    if (doCache) {
        caches[cmd.topic].push(cmd);
    }

    for (const subscriber of subscribers) {
        sendCmd(subscriber, cmd);
    }

    sendSuccess(socket);
}

function handleCommand(cmd, socket) {
    const handlers = {
        subscribe: handleSubscribe,
        unsubscribe: handleUnsubscribe,
        send: handleSend
    };

    const handler = handlers[cmd.command];
    if (handler) {
        handler(cmd, socket);
    }
}

function handleClient(socket) {
    logging.info("New client connected...");

    const rl = readline.createInterface({
        input: socket,
        crlfDelay: Infinity
    });

    socket.on('error', (err) => {
        logging.error('Socket error on connection with', socket.remoteAddress, ':', err);
    });

    rl.on('line', (line) => {
        line = line.trim();
        if (line === "quit") {
            rl.close();
            return;
        }
        if (!line) {
            logging.info("Client disconnected (EOF)");
            rl.close();
            return;
        }
        try {
            line = decoder.write(Buffer.from(line, 'utf8')).trim();
        } catch (e) {
            sendFailure(socket, "Could not decode input as UTF-8");
            return;
        }
        if (line === "") return;
        logging.info(`Received: ${line}`);
        let cmd;
        try {
            cmd = JSON.parse(line);
        } catch (e) {
            sendFailure(socket, "Could not parse json");
            return;
        }
        try {
            if (verifyCommand(cmd)) {
                handleCommand(cmd, socket);
            } else {
                sendFailure(socket, "Malformed json message");
            }
        } catch (e) {
            logging.error(e);
            sendFailure(socket, "Internal exception");
        }
    });

    function cleanupSocket(socket) {
        if (topics_reverse[socket]) {
            for (let topic of topics_reverse[socket]) {
                topics[topic].delete(socket);
            }
            delete topics_reverse[socket];
        }
        logging.info("Client disconnected...");
    }
    
    socket.on('end', () => cleanupSocket(socket));
    socket.on('close', () => cleanupSocket(socket));
}

function runServer(host, port) {
    const server = net.createServer(handleClient);
    server.listen(port, host, () => {
        logging.info(`Listening on ${host}:${port}...`);
    });
}

const DEFAULT_PORT = 7000;
const DEFAULT_CACHE_SIZE = 100;

let port = DEFAULT_PORT;
let cacheSize = DEFAULT_CACHE_SIZE;

if (process.argv.length <= 4) {
    port = parseInt(process.argv[2], 10);
    if (isNaN(port) || port <= 0) {
       port = DEFAULT_PORT;
    }
}
if (process.argv.length == 4) {
    cacheSize = parseInt(process.argv[3], 10);
    if (isNaN(cacheSize) || cacheSize <= 0) {
       cacheSize = DEFAULT_CACHE_SIZE;
    }
}

/* global topics */
let topics = new DefaultDict(() => new Set());
/* global topics_reverse */
let topics_reverse = new DefaultDict(() => new Set());
/* global caches */
let caches = new DefaultDict(() => new Deque(cacheSize));
/* global indexs */
let indexs = new DefaultDict(() => 0);

process.on('unhandledRejection', (reason, promise) => {
    logging.error('Unhandled Rejection at:', promise, 'reason:', reason);
});

runServer('localhost', port);
